### PR TITLE
Add deployment scripts and CI

### PR DESCRIPTION
**Pull request by nmanoogian**
This PR adds deployment scripts and a GitHub Action workflow to deploy all agents in the repo, if necessary.

Closes ENG-4281

### CODE REVIEW

**Comment by nmanoogian** [comment]
Force push removes server-side encryption flags from S3 calls

File: deployment/deploy.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+async function getFileSHA256(filename: string, encoding: crypto.BinaryToTextEncoding) {
+  const fileBuffer = await readFile(filename);
+  const hashSum = createHash("sha256");
+
+  hashSum.update(fileBuffer);
+  return hashSum.digest(encoding);
+}
+
+async function makeTempDir(appName: string) {
+  return await mkdtemp(join(tmpdir(), `secret-agents-build-${appName}-`));
```
Comment (Piccirello) [inline deployment/deploy.ts:51]: nit: is the trailing hyphen intentional?

File: deployment/deploy.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+      },
+    });
+
+    const signingBundleName = `${appPackage.name}-${appPackage.version}.zip`;
+    const signingSourceKey = `unsigned-${signingBundleName}`;
+    const signingPrefix = "signed-";
+
+    const uploadResult = await awsExecJSON<{ VersionId: string }>(
+      `aws s3api put-object --body ${buildOutputPath} --bucket ${config.signingBucket} --key ${signingSourceKey}`
+    );
```
Comment (Piccirello) [inline deployment/deploy.ts:157]: To be safe, we should escape any variables we're using to compose shell commands.

File: deployment/deploy.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    // Write manifest to all buckets
+    await writeFile("manifest.json", JSON.stringify(manifest, null, 2));
+    await Promise.all(
+      config.agentBuckets
+        .slice()
+        .reverse() // Reverse the list to write the primary bucket last
+        .map(async (bucket) => {
+          await awsExec(`aws s3 cp manifest.json s3://${bucket}/manifest.json`);
+        })
+    );
```
Comment (Piccirello) [inline deployment/deploy.ts:233]: Should we handle the case where bucket publishing fails to n buckets (but not all)?

File: deployment/deploy.ts (hunk 1 above)
Comment (nmanoogian) [reply deployment/deploy.ts:51]: Yep, [mkdtemp](https://nodejs.org/api/fs.html#fspromisesmkdtempprefix-options) takes a prefix and appends a random suffix 👍

File: deployment/deploy.ts (hunk 2 above)
Comment (nmanoogian) [reply deployment/deploy.ts:157]: This is a really good point and one that I've definitely been considering. Unfortunately, it doesn't seem that node provides a built-in function for escaping params (though there has been [some discussion](https://github.com/nodejs/node/issues/34840) around it). Like escaping regex, it seems that it's hard to get right given the edge cases. [Stack Overflow has some recommendations](https://stackoverflow.com/questions/1779858/how-do-i-escape-a-string-for-a-shell-command-in-node) but none that really quell my concerns.

[exec](https://nodejs.org/dist/latest-v16.x/docs/api/child_process.html#child_processexeccommand-options-callback) spawns a shell and runs the provided string as a command. Alternatively, some folks have recommended using the [spawn](https://nodejs.org/dist/latest-v16.x/docs/api/child_process.html#child_processspawncommand-args-options) function instead which takes the command and params as discrete values.

What do you think?

Comment (nmanoogian) [reply deployment/deploy.ts]: Update: Spawn works great and feels a lot cleaner and safer; I just pushed that change. Still curious to hear your take though

**Comment by nmanoogian** [comment]
Force push removes accidentally-committed testing code

**Comment by nmanoogian** [comment]
Force push fixes initial manifest creation branch

File: deployment/deploy.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    return [`SpawnError: ${this.message}`, `exit=${this.code}`, this.stderr].join("\n");
+  }
+}
+
+async function spawn(command: string, args: string[], options: SpawnOptions = {}) {
+  console.log(">", command, JSON.stringify(args));
+  return new Promise<{ stdout: string; stderr: string }>((resolve, reject) => {
+    let stdout = "";
+    let stderr = "";
+    const process = spawnEE(command, args, options);
```
Comment (Piccirello) [inline deployment/deploy.ts:81]: From the Node [docs](https://nodejs.org/docs/latest-v16.x/api/child_process.html#child_processspawncommand-args-options):

> If the shell option is enabled, do not pass unsanitized user input to this function. Any input containing shell metacharacters may be used to trigger arbitrary command execution.

Could you explicitly set `options.shell` to `false` with a commenting noting why we're doing so? No callers override `options.shell` today but we should prevent future us from making that mistake.

File: deployment/deploy.ts (hunk 2 above)
Comment (Piccirello) [reply deployment/deploy.ts]: Switching to spawn was a great change - will make sure to keep this in my Node toolbox. I added one comment below around hardening our use of `spawn`.

File: deployment/deploy.ts (hunk 4 above)
Comment (nmanoogian) [reply deployment/deploy.ts:81]: Good callout 👍 I'll add that

**Pull request merged**